core/services/log/registrations.go

```diff
@@ -0,0 +1,271 @@
+package log
+
+import (
+	"sync"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/smartcontractkit/chainlink/core/logger"
+	"github.com/smartcontractkit/chainlink/core/store/models"
+)
+
+// How it works in general:
+// 1. All received logs are kept in a data structure and deleted ONLY after they are too old for all subscribers
```

Comment:
What happens if the node crashes or is restarted?

Reply:
What determines if a log is "too old"?

Reply:
too old == when given log height is lower than (latest height - highest num_confirmations) - added it to the comment

"What happens if the node crashes or is restarted?" - logs will be backfilled from the latest head from DB, for Subscribers that are added before all dependents of LogBroadcaster are done

If a subscriber is added after the LogBroadcaster does the initial backfill, then it's possible/likely that the backfill fill only have depth:1 (from latest head)

Of course, these backfilled logs + any new logs will only be sent after the num_confirmations for given subscriber.

I'll add that to the comment